### PR TITLE
adjust tests to use dedicated path.data per node

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Follow up to https://github.com/crate/crate-qa/pull/214 (which had semantical issue of having single path.data)
Fixes https://github.com/crate/crate-alerts/issues/249

this change reflects idea from https://github.com/elastic/elasticsearch/commit/770d8e9e39c78bc1e7d7521a7782194eaab36624:
keep settings per node which we already do in 
https://github.com/crate/crate-qa/blob/master/src/crate/qa/tests.py#L262
and re-use path.data setting if a test case needs to start cluster more than one time.

related to setting removal: relevant `InternalTestCluster` changes were backported in https://github.com/crate/crate/pull/8790 and https://github.com/crate/crate/pull/8807

## Checklist

 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

**Comment by BaurzhanSakhariev** [comment]
I checked that all usages of the new_cluster followed by another call(s) handled by copying data_path from the first call and used further.

in some cases, like HotfixDowngradeTest we use only upgrade_node, which sends old node's settings and we re-use them on a node creation.

File: tests/snapshot/test_snapshot.py [hunk 1]
```diff
@@ -32,53 +27,54 @@ def test_snapshot_restore_and_drop_in_parallel(self):
...
+                    try:
+                        c.execute('DROP SNAPSHOT repo.snapshot2')
+                    except ProgrammingError:
+                        self.fail("Restore and Drop Snapshot operation should work in parallel")
+
+                    assert_busy(lambda: self._assert_num_docs(conn, num_docs))
+
+                cluster.stop()
+        finally:
+            self.tearDown()
```
Comment (BaurzhanSakhariev) [inline tests/snapshot/test_snapshot.py:80]: we used to use one specific folder  and then clean it up with `shutil.rmtree(self.PATH_DATA, ignore_errors=True)`
now each node has it's own folder which cleanup on teardown

File: src/crate/qa/tests.py [hunk 2]
```diff
@@ -181,7 +184,8 @@ def _unicast_hosts(self, num, transport_port=4300):
             for x in range(num)
         ])
 
-    def _new_cluster(self, version, num_nodes, settings=None, env=None):
+    def _new_cluster(self, version, num_nodes, data_paths=None, settings=None, env=None):
```
Comment (BaurzhanSakhariev) [inline src/crate/qa/tests.py:187]: there is also `_new_heterogeneous_cluster` - but I didn't change it as it's not used (usages was removed in https://github.com/crate/crate-qa/commit/10bc51176ff71e6015d9bdc91b56781c4fe5c7a4 and https://github.com/crate/crate-qa/commit/66b2b68879fb1e323b94fe7da10e5a90942be43d). 
Should I remove it in a separate commit?

**Comment by BaurzhanSakhariev** [comment]
Hi @seut, could you please take a look? Every commit fixes different test failure.

File: src/crate/qa/tests.py [hunk 3]
```diff
@@ -228,16 +237,22 @@ def setUp(self):
         self._log_consumers = []
 
         def new_node(version, settings=None, env=None):
-            self._path_data = self.mkdtemp()
```
Comment (seut) [inline src/crate/qa/tests.py:231]: Why don't we just turn `self._path_data` in to a list and store each nodes generated temp path inside? this way, `tearDown` functionality would still work

File: tests/bwc/test_upgrade.py [hunk 4]
```diff
@@ -471,6 +480,13 @@ class SnapshotCompatibilityTest(NodeProvider, unittest.TestCase):
     VERSION = ('3.3.x', '4.x.x')
 
     def test_snapshot_compatibility(self):
+        try:
+            self.setUp()
```
Comment (seut) [inline tests/bwc/test_upgrade.py:484]: why is this needed? `setUp` and `tearDown` should be called by pythons unittest runner

File: tests/snapshot/test_snapshot.py [hunk 5]
```diff
@@ -12,11 +11,6 @@
 
 class SnapshotOperationTest(NodeProvider, unittest.TestCase):
 
-    PATH_DATA = 'data_test_snapshot_ops'
-
-    def tearDown(self):
```
Comment (seut) [inline tests/snapshot/test_snapshot.py:17]: see my comment above, would be great to keep that.

File: src/crate/qa/tests.py (hunk 3 above)
Comment (BaurzhanSakhariev) [reply src/crate/qa/tests.py:231]: afaik, tearDown [removes](https://github.com/crate/crate-qa/blob/master/src/crate/qa/tests.py#L271) tmp dir's  created by `self.mkdtemp()`

File: tests/snapshot/test_snapshot.py (hunk 5 above)
Comment (BaurzhanSakhariev) [reply tests/snapshot/test_snapshot.py:17]: I couldn't use it as it's removes specific single folder. I added `tearDown` of the base class instead in finally block as it takes care about folders created by `mkdtemp`

File: src/crate/qa/tests.py (hunk 3 above)
Comment (seut) [reply src/crate/qa/tests.py:231]: ah right, even better

File: tests/snapshot/test_snapshot.py [hunk 6]
```diff
@@ -32,53 +26,54 @@ def test_snapshot_restore_and_drop_in_parallel(self):
...
+                    try:
+                        c.execute('DROP SNAPSHOT repo.snapshot2')
+                    except ProgrammingError:
+                        self.fail("Restore and Drop Snapshot operation should work in parallel")
+
+                    assert_busy(lambda: self._assert_num_docs(conn, num_docs))
+
+                cluster.stop()
+        finally:
+            self.tearDown()
```
Comment (seut) [inline tests/snapshot/test_snapshot.py:79]: self.tearDown() should be called automatically, why do we need to call it explicitly?

File: tests/bwc/test_upgrade.py (hunk 4 above)
Comment (BaurzhanSakhariev) [reply tests/bwc/test_upgrade.py]: initially I added it just by following [example](https://github.com/crate/crate-qa/pull/216/commits/dc0c9fdd211b0b9a7bbad59a8c6d87f6320594de) from another test case

File: tests/snapshot/test_snapshot.py (hunk 5 above)
Comment (BaurzhanSakhariev) [reply tests/snapshot/test_snapshot.py:17]: removed teardown from finally - thx for the hint about automatic call

File: tests/bwc/test_upgrade.py (hunk 4 above)
Comment (BaurzhanSakhariev) [reply tests/bwc/test_upgrade.py]: ok, [code](https://github.com/crate/crate-qa/blob/master/tests/bwc/test_upgrade.py#L154-L160)  I taken as example explicitly calls setUp/tearDown as it runs this test for series of upgrade_paths and needs to refresh things in between - so I won't change it

**Review by seut — APPROVED** [review]
👍

**Comment by BaurzhanSakhariev** [comment]
retest this please (stuck on test_snapshot_compatibility)

**Comment by BaurzhanSakhariev** [comment]
retest this please (stuck on test_snapshot_compatibility)
